src/numeric-value-typing.js

```diff
@@ -0,0 +1,282 @@
+const failure = null;
+const baseTypes = ["percent", "length", "angle", "time", "frequency", "resolution", "flex"];
+
+const compatibleUnitGroups = [{
+  // Absolute lengths
+  units: ["cm", "mm", "Q", "in", "pt", "pc", "px"], canonicalUnit: "px", ratios: {
+    "cm": 96 / 2.54, "mm": (96 / 2.54) / 10, "Q": (96 / 2.54) / 40, "in": 96, "pc": 96 / 6, "pt": 96 / 72, "px": 1
+  }
+}, {
+  // Angle
+  units: ["deg", "grad", "rad", "turn"], canonicalUnit: "deg", ratios: {
+    "deg": 1, "grad": 360 / 400, "rad": 180 / Math.PI, "turn": 360
+  }
+}, {
+  // Time
+  units: ["s", "ms"], canonicalUnit: "s", ratios: {
+    "s": 1, "ms": 1 / 1000
+  }
+}, {
+  // Frequency
+  units: ["hz", "khz"], canonicalUnit: "hz", ratios: {
+    "hz": 1, "khz": 1000
+  }
+}, {
+  // Resolution
+  units: ["dpi", "dpcm", "dppx"], canonicalUnit: "dppx", ratios: {
+    "dpi": 1 / 96, "dpcm": 2.54 / 96, "dppx": 1
+  }
+}];
+
+const unitToCompatibleUnitsMap = new Map();
+for (const group of compatibleUnitGroups) {
+  for (const unit of group.units) {
+    unitToCompatibleUnitsMap.set(unit, group);
+  }
+}
+
+export function getSetOfCompatibleUnits(unit) {
+  return unitToCompatibleUnitsMap.get(unit);
+}
+
+function productOfUnionMaps(units1, units2) {
+  // 1. Let result be a copy of units1.
+  const result = {...units1};
+  // 2. For each unit → power in units2:
+  for (const unit of Object.keys(units2)) {
+    if (result[unit]) {
+      // 1. If result[unit] exists, increment result[unit] by power.
+      result[unit] += units2[unit];
+    } else {
+      // 2. Otherwise, set result[unit] to power.
+      result[unit] = units2[unit];
+    }
+  }
+  // 3. Return result.
+  return result;
+}
+
+export function createSumValue(self) {
```

Comment:
The parameter name 'self' seems a bit unusual, since it is normally used to indicate a context.  How about calling the parameter cssNumericValue.

Reply:
Sounds good. Renamed the parameter 👍